spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/NetChangelogIterator.java

```diff
@@ -0,0 +1,138 @@
+   * data. For example, if there are two identical records, it will remove one of them.
+   */
+  @Override
+  public Row next() {
+    // return the cached row if there is one
+    if (postImage != null) {
+      Row currentRow = postImage;
+      postImage = null;
+      return currentRow;
+    }
+
+    // return the first row if there is no more rows
+    if (!getRowIterator().hasNext() && firstRow != null) {
+      Row currentRow = firstRow;
+      firstRow = null;
+      return currentRow;
+    }
+
+    if (firstRow == null) {
+      firstRow = getRowIterator().next();
+      return null;
+    }
+
+    Row lastRow = null;
+
+    while (getRowIterator().hasNext()) {
+      Row currentRow = getRowIterator().next();
+      // if current row is not the same logical as the first row
+      if (sameLogicalRow(firstRow, currentRow)) {
+        lastRow = currentRow;
+      } else {
+        // if current row is not the same logical row as the first row, compute the results and
+        // set the current row as the first row, and return the result
+        Row result = getNetChanges(firstRow, lastRow);
+        firstRow = currentRow;
+        return result;
+      }
+    }
+
+    Row current = getNetChanges(firstRow, lastRow);
+    firstRow = null;
+    return current;
+  }
+
+  /**
+   * This method gets the net changes between the first version and the last version of a logical
+   * row. Here are the algorithm:
+   *
+   * <ol>
+   *   <li>If the first version is an insert, and last one is an insert, then return an insert with
+   *       the last version.
+   *   <li>If the first version is an insert, and last one is a delete, then return null.
+   *   <li>If the first version is a delete, and last one is a delete, then return a delete.
+   *   <li>If the first version is a delete, and last one is an insert, if they are the exactly
+   *       same, they are carry-over rows, return null, otherwise return an update pair.
+   * </ol>
+   */
+  private Row getNetChanges(Row first, Row last) {
+    if (first == last || last == null) {
+      return first;
+    }
+
+    if (first.getString(getChangeTypeIndex()).equals(INSERT)) {
+      if (last.getString(getChangeTypeIndex()).equals(INSERT)) {
+        return last;
+      }
+
+      if (last.getString(getChangeTypeIndex()).equals(DELETE)) {
+        return null;
+      }
+    }
+
+    if (first.getString(getChangeTypeIndex()).equals(DELETE)) {
+      if (last.getString(getChangeTypeIndex()).equals(DELETE)) {
+        return first;
+      }
+
+      if (last.getString(getChangeTypeIndex()).equals(INSERT)) {
+        if (isCarryoverRecord(first, last)) {
+          return null;
+        } else {
+          // update
+          Row updatedFirstRow = modify(first, getChangeTypeIndex(), UPDATE_BEFORE);
+          postImage = modify(last, getChangeTypeIndex(), UPDATE_AFTER);
+          return updatedFirstRow;
+        }
+      }
+    }
+
+    return null;
```

Comment:
What condition is this?

Reply:
A row with change type is neither `DELETE` nor `INSERT`. It shouldn't happen. I think we can throw an exception here. Will make the change in the next commit.